File: Data Structures & Algorithms/binary-tree-diameter/submission-2.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {
    //diameter of a binary tree is the number of edges on the longest path 
    //so use DFS 
    static int maxDiameter = 0;

    private int DfsHelper(TreeNode node){
        if(node == null) return 0; 

        int leftH = DfsHelper(node.left);
        int rightH = DfsHelper(node.right);

        if(leftH + rightH > maxDiameter){
            maxDiameter = leftH + rightH;
        }

        return 1 + Math.Max(leftH, rightH);
    }
    public int DiameterOfBinaryTree(TreeNode root) {
        maxDiameter = 0;
        DfsHelper(root);

        return maxDiameter;
    }
}

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

//do inorder traversal 
//compare root node to max (integer min value), increment counter by one 
//then set max value to root, redo for left child
//if left child greater than root, increment counter, otherwise continue 
//then compare to rigth child and do the same

class Solution {
    private int count = 0; 

    private void findGood(TreeNode node, int max){
        if(node == null) return; 

        if(node.val >= max){
            count++;

            max = node.val; 
        }

        findGood(node.left, max);
        findGood(node.right, max);
        
    }

    public int goodNodes(TreeNode root) {
        
        findGood(root, Integer.MIN_VALUE); 

        return count; 
    }
}
